Zoom, Rendering Notification, and Misc Cleanup

DXCanas opened the pull request:
Solves the console errors described in https://github.com/learningequality/kolibri/pull/1888#pullrequestreview-51858828

DXCanas committed:
Initializing progress fix <progress-bar>

DXCanas committed:
Using computed prop for loading conditional
renaming displayPage to getPageViewport
Adding name to vue component

DXCanas committed:
Moving pdf loading progress to `mounted()` because doing so in `getDocument()` was deprecated and to minimize `<progress-bar>` flicker.
Moved data retrieval/setting to `mounted()` to consolidate `getDocument()` `.then`s

DXCanas committed:
Debouncing scroll-event-driven to prevent superfluous de/re-renders

DXCanas committed:
Setting up initial scale of page according to screensize
Including second parameter in getPage() as suggested by Richard. May remove.

DXCanas committed:
Fixing fullscreen button error - text would not update if you exited fullscreen by any means other than clicking the button.
Had to update the dependency for this fix.

DXCanas committed:
Clearing out pdfObject, as it's no longer used in this plugin

DXCanas committed:
Rough Draft Zoom

DXCanas commented on kolibri/plugins/document_pdf_render/assets/src/views/index.vue:
```diff
@@ -13,8 +13,12 @@
       <mat-svg v-else class="icon" category="navigation" name="fullscreen"/>
     </icon-button>
+
+    <icon-button text="Zoom In" @click="zoomIn()"/>
```
Unwrapped, unstyled. WIll be soon!

DXCanas commented on kolibri/plugins/document_pdf_render/assets/src/views/index.vue:
```diff
@@ -13,8 +13,12 @@
+
     <div ref="pdfcontainer" class="pdfcontainer" @scroll="checkPages">
-      <progress-bar v-if="loading" class="progress-bar" :show-percentage="true" :progress="progress"/>
+      <progress-bar v-if="documentLoading" class="progress-bar" :show-percentage="true" :progress="progress"/>
```
Just using a computed prop here. Thinking about going back to manually setting to have the page rendered on arrival

DXCanas commented on kolibri/plugins/document_pdf_render/assets/src/views/index.vue:
```diff
@@ -63,48 +75,83 @@
-            ScreenFull.toggle(this.$refs.container);
-          }
-          this.isFullscreen = false;
+        if (this.fullscreenAllowed) {
```
A minor fullscreen bugfix. Might be best to break this into a separate PR :man_shrugging:

DXCanas commented on kolibri/plugins/document_pdf_render/assets/src/views/index.vue:
```diff
@@ -63,48 +75,83 @@
+        this.scale -= 0.1;
       },
-      displayPage(pdfPage) {
+      getPage(pageNum, firstRender = false) {
```
Probably ditching the firstRenderOption? Or just renaming `setupIntialPageScale` to `setupZoom`

DXCanas commented on kolibri/plugins/document_pdf_render/assets/src/views/index.vue:
```diff
@@ -63,48 +75,83 @@
+          this.scale = this.elSize.height / pdfPageHeight;
+        }
+      },
+      startRender(pdfPage) {
```
Renamed it `startRender` to avoid confusion between `display` and `show`

DXCanas commented on kolibri/plugins/document_pdf_render/assets/src/views/index.vue:
```diff
@@ -175,49 +220,69 @@
       },
-      checkPages() {
+      // debouncing so we're not de/re-render many pages unnecessarily
+      checkPages: debounce(function() {
```
Added the debounce here

DXCanas commented on kolibri/plugins/document_pdf_render/assets/src/views/index.vue:
```diff
@@ -175,49 +220,69 @@
           }
         }
-      },
+      }, renderDebounceTime),
```
Set outside of the `exports` scope

DXCanas commented on kolibri/plugins/document_pdf_render/assets/src/views/index.vue:
```diff
@@ -175,49 +220,69 @@
     },
     watch: {
       scrollPos: 'checkPages',
+      scale(newScale, oldScale) {
```
The `scale` watcher

DXCanas commented on kolibri/plugins/document_pdf_render/assets/src/views/index.vue:
```diff
@@ -175,49 +220,69 @@
-        // Begin retrieving the first page
-        return this.getPage(1);
-      });
+      this.loadPdfPromise = PDFJSLib.getDocument(this.defaultFile.storage_url);
```
Stripped created down to be a little simpler. Moved a lot of the `.then` logic down into mounted. Haven't noticed any issues, but please advise

DXCanas commented on kolibri/plugins/document_pdf_render/assets/src/views/index.vue:
```diff
@@ -175,49 +220,69 @@
-      this.pdfloadingPromise.then(() => {
-        this.showPage(1);
-        this.$emit('startTracking');
+      if (this.fullscreenAllowed) {
```
A part of the fullscreen bugfix. It wasn't updating the text/icon if you used the escape button

DXCanas commented on kolibri/plugins/document_pdf_render/assets/src/views/index.vue:
```diff
@@ -175,49 +220,69 @@
+        });
+      }
+
+      // pass callback to update loading bar
```
This is the "supported" way to specify a progress callback

rtibbles replied:
I think you should add this within the `created` handler, rather than `mounted` - data is available to be modified, it just won't be used for rendering yet.

DXCanas commented on kolibri/plugins/document_pdf_render/assets/src/views/index.vue:
```diff
@@ -175,49 +220,69 @@
+
+        // Retrieve the first Page object
+        this.getPage(1).then(pdfPage => {
+          this.setupInitialPageScale(pdfPage);
```
Where the page scale is specified

DXCanas commented on package.json:
```diff
@@ -46,7 +46,7 @@
     "purecss": "^0.6.2",
     "rest": "1.3.2",
-    "screenfull": "3.0.2",
+    "screenfull": "3.3.1",
```
Tested a bit, not noticing any issues

DXCanas commented on kolibri/plugins/document_pdf_render/assets/src/views/index.vue:
```diff
@@ -63,48 +75,83 @@
-            const ctx = canvas.getContext('2d');
+
+            // define canvas and dummy blank page dimensions
+            canvas.width = this.pageWidth = viewport.width;
```
both canvas and pageWidth are being defined all the time now.

DXCanas reviewed (commented):
Quick comments on some changes I've made!

rtibbles commented on kolibri/plugins/document_pdf_render/assets/src/views/index.vue:
```diff
@@ -175,49 +219,69 @@
+        this.pdfPages = {};
+
+        // Retrieve the first Page object
+        this.getPage(1).then(pdfPage => {
```
Note that you are now deferring getting the first page until the component has mounted - this should be moved back into the `created` handler, so that we are not waiting on DOM mounting for this operation. The only thing we need to wait for DOM mounting for is actually rendering into the page.

DXCanas replied:
I was experiencing a flicker with the loading bar, and thought it might be because it'd been loading prior to being mounted. Thought moving it to mounted would fix that, but I'm not seeing any huge benefit

rtibbles commented:
On tablet landscape mode, I think I would expect the PDF to expand horizontally more by default?
![image](https://user-images.githubusercontent.com/1680573/28835588-bd2c0d14-769a-11e7-9c08-de82a5a159fe.png)

Could be a bit wider? And getting overflow of the Author field below in tablet portrait mode:
![image](https://user-images.githubusercontent.com/1680573/28835643-f7d3413a-769a-11e7-967c-a0e937f98eef.png)


Look out for button overlap on mobile portrait:
![image](https://user-images.githubusercontent.com/1680573/28835606-d2ef762c-769a-11e7-88df-c967258fccf4.png)

rtibbles reviewed (changes requested):
Renders smoothly, and with far fewer warnings about request animation frame delays, so I think the debouncing is working well.

Couple of comments here, plus some UI comments with screenshots.

DXCanas commented:
> On tablet landscape mode, I think I would expect the PDF to expand horizontally more by default?

So it'd start of rendered at full width on both tablet and phone? (landscape and portrait)

Do we just want to avoid full page height all together? I took this approach due to feedback from khang

> Could be a bit wider?

I agree, though @jessica mentioned that it got annoying for users to scroll the page when the pdf was taking up so much width. They had nowhere to grab the page itself.

> Look out for button overlap on mobile portrait:

Forgot to mention that they're going to be fixed to the bottom right. Jessica thought just icons would be best. Do you have opinions on that?

rtibbles commented:
It's just a lot of grey space for a very small area of content. I can see the advantage of allowing people to see a page at a time, and we would want to avoid using the full width on ultra wide screens, but I think on 4:3 aspect ratio we should probably be using the full width.

On mobile users should be able to press on the page itself to scroll, and as long as we are not going so wide as to push out the scroll bar, should be fine on desktop?

No strong opinions on the buttons, bottom right display works for me, consistent with Chrome built in PDF display.

@khangmach @jtamiace input appreciated.

khangmach commented:
Can we have the background of the pdf container be something darker than this current #f9f9f9? Maybe #212121? Having a background gives something to compare the width of the PDF to and makes margins look more visible. Buttons bottom right sound fine to me.

jtamiace commented:
**Mobile**
I think there needs to be enough space on either side of the PDF renderer in order to allow the user to scroll vertically with their finger, so I'm okay with the grey space on the sides. Reading a PDF on mobile is going to be a painful experience no matter what, and there is the fullscreen mode that they can enable. 

Agree that the PDF itself should be rendered full-width also.
Also, using icon buttons for zoom in/out goes against the style guide but I think this is a very reasonable exception. Bottom right sounds fine here because you can see the entire renderer.

**Tablet**
I think it's also reasonable for the PDF to render full-width, both landscape and portrait.

Icon buttons might be good to align top/right at this breakpoint since the PDF renderer may extend offscreen at the bottom and hide the buttons.

**Desktop**
I don't think it's a good idea to render full-width OR fit to the height of the screen. The document is unreadable when it's that small, and also when it's too big. Would suggest that the document by default loads to some size in between. I don't think the user loses much by not seeing 100% of every page.

Icon buttons might be good to align top/right at this breakpoint since the PDF renderer may extend offscreen at the bottom and hide the buttons.

jtamiace commented:
Regarding background color, @DXCanas did you remove the dark grey that you showed me yesterday? It looked nice

rtibbles commented:
For desktop, we can just default to the document scale of `1.0`?

@jtamiace get your point about scrolling now, it's about scrolling the Kolibri page, not the PDF page.

rtibbles commented:
I think no matter where they zoom buttons are placed, they should be fixed inside the PDF container and kept visible. I think the intention is for the height of the container itself not to increase beyond the height of the screen so they should be kept visible?

jtamiace commented:
>  I think the intention is for the height of the container itself not to increase beyond the height of the screen so they should be kept visible?

If it is possible for the container to be scaled to browser height, bottom-right would work for consistency. It just looks awkward on desktop currently, where the container takes up about 50-60% of the available screen height.

<img width="1630" alt="screen shot 2017-08-01 at 10 16 23 am" src="https://user-images.githubusercontent.com/6668144/28837780-daa4d058-76a2-11e7-8eb8-318242485c41.png">

DXCanas committed:
Adding nicer zoom buttons and styles.
Including core theme to have colors available.
Adding an error flag for future use.
Moving non-render activity from `mounted` to `created`.
Removing parameter from `setupInitialScale`.
Removing optional `firstRender` from `getPage`.

DXCanas commented:
@jtamiace @khangmach this is all still WIP. Was mostly for code review. I'll get all the suggestions in by the end

DXCanas commented:
So, I've got:
- [x] Darker Background. Might wait on this until I merge latest develop so that I get the new colors(?)
- [x] Make it a bit taller but not much.
    - @jtamiace Calculating how much of the display we have while taking into account the nav, tabs, and breadcrumbs is difficult because they take up a different proportion of the screen depending on the height of the display. They're also not defined proportionally, they're defined using pixels, Right now, they're being guesstimated - I can guesstimate some more
- [x] PDF is full-width in every scenario BUT full size desktop, where it's just rendered to scale of the page
    - Full height is never a thing
- [X] Align zoom buttons to bottom on mobile
- [x] Align zoom button to top on desktop (unless you can find the magic height that's not awk and doesn't ever fall off the fold
- [x] Make sure pages to leave the container and cover the author
- [ ] Loading Message

Am I missing anything?

DXCanas commented:
Update: Currently, there's a min height on the pdf viewer's box, so landscape tablet does go below the fold. 

Removing it makes it look ridiculous, though:
![image](https://user-images.githubusercontent.com/9877852/28839566-e6e0b264-76a8-11e7-81b8-2abb7b33db9a.png)

Gonna go with top-align in these cases

DXCanas committed:
Defaulting to full width, except in cases with a large display (Desktop).
Adjusting height of the pdfWrapper, now closer to the fold

DXCanas committed:
Conditionally aligning buttons to top of the viewer, rather than bottom.

jtamiace commented:
Might help the buttons pop out a bit more if you add drop shadow? Same one that appears on the raised buttons

DXCanas commented:
@jtamiace I had issues adding a drop shadow to keen's UI buttons, but I'll give it a shot.

DXCanas committed:
Adding background color
Using <section> rather than <p> for page containers.
Adding some margins to left and right of pages.

DXCanas committed:
Fixing race condition that caused all pages to be rendered.
Prior to encountering this condition, handled by only rendering first page and no other.
Also clearing out errors.

DXCanas committed:
First pass loading message.
Wrapping page number string.
Also cleaning up some css classes.

DXCanas commented:
OK! I think it's ready for review

DXCanas commented:
![pdfdemo](https://user-images.githubusercontent.com/9877852/28900933-90e23bc4-77a9-11e7-8cae-b5541084ccc3.gif)
@jtamiace @khangmach 

UI buttons don't lend themselves well to drop shadows, that's why they're not there @jtamiace
